### PR TITLE
Add support for STM32G4 family

### PR DESCRIPTION
**Pull request by rleh**
- [x] Blink example compiles! :wink: 
- [x] Adapt peripherals
  - [x] ADC
  - [x] ~~DAC~~ There is generally no DAC driver yet.
  - [x] Comparator
  - [x] ~~**New** FDCAN driver~~ -> different PR
- [x] Add [Nucleo-G474RE](https://de.farnell.com/stmicroelectronics/nucleo-g474re/entwicklungsboard-32bit-arm-cortex/dp/3132424?CMP=GRHB-OCTOPART) BSP and examples
- [x] Test on real hardware

### CODE REVIEW

**Comment by rleh** [comment]
There are a lot of this _redefined_ errors:
```
modm/src/modm/platform/uart/uart_base.hpp:25: warning: "USART_ISR_RXNE" redefined
   25 | #define USART_ISR_RXNE    USART_ISR_RXNE_RXFNE
      | 
In file included from modm/src/modm/platform/clock/../device.hpp:25,
[...]
modm/ext/cmsis/device/stm32g474xx.h:15998: note: this is the location of the previous definition
15998 | #define USART_ISR_RXNE               USART_ISR_RXNE_Msk                        /*!< Read Data Register Not Empty */
```
Another ST inconsistency?

File: ext/st/device.hpp.in [hunk 1]
```diff
@@ -29,6 +29,8 @@
 #include <{{ header }}>
 %% endfor
 
+#define HRTIM_TIM_TypeDef HRTIM_TypeDef
```
Comment (rleh) [inline ext/st/device.hpp.in]: Uuuuugly hack :laughing: 

I don't get where modm generates `HRTIM_TIM_TypeDef` from.

---
```
modm/src/modm/platform/clock/../device.hpp:119:8: error: 'HRTIM_TIM_TypeDef' does not name a type; did you mean 'HRTIM_TypeDef'?
```

Comment (salkinium) [reply ext/st/device.hpp.in]: Implementation of this https://modm.io/reference/module/modm-cmsis-device/#debugging-peripherals

Comment (rleh) [reply ext/st/device.hpp.in]: Yes, but whats makes modm think of `HRTIM_TIM_TypeDef` instead of `HRTIM_TypeDef` which would be correct?

Comment (salkinium) [reply ext/st/device.hpp.in]: The description is wrong, the module only declares extern objects to make the debugger aware of the types. Hacky hacky.

Comment (salkinium) [reply ext/st/device.hpp.in]: It regexes that out of the CMSIS Header files, so ST probably messed up again.

Comment (salkinium) [reply ext/st/device.hpp.in]: This regex to be precise: https://github.com/modm-io/modm/blob/develop/ext/st/module.lb#L87

**Comment by salkinium** [comment]
Please update the STM32 CMSIS Header submodule to the latest commit, so that you're not working with bugs that have already been fixed!

**Comment by salkinium** [comment]
ST already fixed their own header files:
```c
#define USART_CR1_TXEIE_Pos          (7U)
#define USART_CR1_TXEIE_Msk          (0x1UL << USART_CR1_TXEIE_Pos)            /*!< 0x00000080 */
#define USART_CR1_TXEIE              USART_CR1_TXEIE_Msk                       /*!< TXE Interrupt Enable */
#define USART_CR1_TXEIE_TXFNFIE_Pos  USART_CR1_TXEIE_Pos
#define USART_CR1_TXEIE_TXFNFIE_Msk  USART_CR1_TXEIE_Msk                       /*!< 0x00000080 */
#define USART_CR1_TXEIE_TXFNFIE      USART_CR1_TXEIE_Msk                       /*!< TXE and TX FIFO Not Full Interrupt Enable */
```
and
```c

#define USART_ISR_TXE_Pos            (7U)
#define USART_ISR_TXE_Msk            (0x1UL << USART_ISR_TXE_Pos)              /*!< 0x00000080 */
#define USART_ISR_TXE                USART_ISR_TXE_Msk                         /*!< Transmit Data Register Empty */
#define USART_ISR_TXE_TXFNF_Pos      USART_ISR_TXE_Pos
#define USART_ISR_TXE_TXFNF_Msk      USART_ISR_TXE_Msk                       /*!< 0x00000080 */
#define USART_ISR_TXE_TXFNF          USART_ISR_TXE_Msk                       /*!< Transmit Data Register Empty or TX FIFO Not Full Flag */
```

So the lines in `uart_base.hpp` requires a few `#ifndef`.

**Comment by salkinium** [comment]
See my branch here: https://github.com/salkinium/modm/tree/feature/stm32g4

**Comment by rleh** [comment]
The *build-docs* CI job fails:
```
...
Merging module tree...
/bin/bash: line 1:   405 Killed                  python3 tools/scripts/generate_module_docs.py
Exited with code 137

Hint: Exit code 137 typically means the process is killed because it was running out of memory
Hint: Check if you can optimize the memory usage in your app
Hint: Max memory usage of this container is 4283166720
 according to /sys/fs/cgroup/memory/memory.max_usage_in_bytes
```

**Comment by salkinium** [comment]
Try this patch, it should only use 1.5GB of memory by merging the module tree immediately instead of storing all trees and then merging (which was stupid).
https://gist.github.com/salkinium/4cd2509cb2bf8d28ac2c022320fa11f5

**Comment by rleh** [comment]
There is a issue with OpenOCD: The latest OpenOCD Version (0.10.0 from January 2017) neither supports the STM32G4 as a target, nor the ST-Link V3 on the Nucleo-G474RE board.

There is probably no way out to use a newer OpenOCD version than 0.10.0. So compile OpenOCD yourself or use the AUR version (Git HEAD) for Arch-Linux.

To fix the missing STM32G4 support I will try to add the necessary files in `tools/openocd/modm/`.

StLink-V3 support: http://openocd.zylin.com/#/c/4717/
STM32G4 support: http://openocd.zylin.com/#/c/4807/

**Comment by salkinium** [comment]
OpenOCD also has no STM32G0 support either, I've been using the ST-Link mass storage device to flash it, but debugging would be useful too. OpenOCD has a very slow PR process, as in like multiple years.

File: src/modm/platform/comp/stm32/base.hpp.in [hunk 2]
```diff
@@ -58,6 +62,28 @@ namespace modm::platform
 		};
 	protected:
 		static constexpr uint32_t HysteresisMask = 0b11 << 16;
+	{% elif driver.type in ["stm32-tsmc90_g4_rockfish_cube"] -%}
```
Comment (salkinium) [inline src/modm/platform/comp/stm32/base.hpp.in]: 1. Why not `%%`?
2. `elif "g4" in driver.type`? Probably less fragile.

File: src/modm/platform/comp/stm32/module.lb [hunk 3]
```diff
@@ -48,7 +68,22 @@ def prepare(module, options):
...
+        "stm32-tsmc90_orca128_cube"
+        "stm32-tsmc90_orca512_cube"
+        "stm32-tsmc90_orcazero_cube"
+        "stm32-v1.0"
+        "stm32-v1.2"
+        "stm32-v1.3"
+        "stm32-v3.4"
+        "stm32-v3.6"
+    """
+    if not device.get_driver("comp")["type"] in ["stm32-v1.3", "stm32-tsmc90_cube", "stm32-tsmc90_g4_rockfish_cube"]:
```
Comment (salkinium) [inline src/modm/platform/comp/stm32/module.lb:86]: Hm.

File: ext/st/device.hpp.in (hunk 1 above)
Comment (salkinium) [reply ext/st/device.hpp.in]: Is this still an issue with updated cmsis-header submodule?

**Review by salkinium — COMMENTED** [review]
Tested in hardware?

File: src/modm/platform/comp/stm32/module.lb (hunk 3 above)
Comment (rleh) [reply src/modm/platform/comp/stm32/module.lb:86]: Better idea?

File: ext/st/device.hpp.in (hunk 1 above)
Comment (rleh) [reply ext/st/device.hpp.in]: Let me check that...

File: src/modm/platform/comp/stm32/module.lb (hunk 3 above)
Comment (salkinium) [reply src/modm/platform/comp/stm32/module.lb:86]: Rename it in modm-devices, but I just made it incompatible with the latest identifier extension, so hm. Guess we'll fix it later.

File: src/modm/platform/comp/stm32/base.hpp.in (hunk 2 above)
Comment (rleh) [reply src/modm/platform/comp/stm32/base.hpp.in]: > 1. Why not `%%`?

Consistency...

> 2. `elif "g4" in driver.type`? Probably less fragile.

Fixed!

File: ext/st/device.hpp.in (hunk 1 above)
Comment (rleh) [reply ext/st/device.hpp.in]: Nope.

**Comment by rleh** [comment]
Test on real hardware: ✅

Merge?

**Comment by salkinium** [comment]
Can you add the family to the README, and check if there're any other docs that need updating?

**Comment by salkinium** [comment]
Is the linkerscript chosen correctly?

**Comment by rleh** [comment]
> Is the linkerscript chosen correctly?

The ICCM Linkerscript is chosen, looks right to me.
[STM32G474 Memory](https://www.st.com/resource/en/datasheet/stm32g474cb.pdf)

**Comment by rleh** [comment]
> Can you add the family to the README, and check if there're any other docs that need updating?

README done.
I cannot find any other documents that need to be updated.

**Review by salkinium — APPROVED** [review]
Fantastic!

**Comment by rleh** [comment]
I wrote a small blog-post about how to compile OpenOCD with support for STM32G0 and STM32G4:
https://rleh.de/2019/10/08/openocd-stm32-stm32g4-stm32g0.html

**Pull request merged**